cmake: kconfig: Fix rerunning cmake after Kconfig warnings

ulfalizer opened the pull request:
Commit b3d165f ("scripts: kconfig: Handle warnings generated
during evaluation") made it common for kconfig.py to fail after writing
zephyr/.config. This confuses the configuration fragment checksum logic
in cmake/kconfig.cmake, because it expects the saved checksum file to
exist if zephyr/.config exists.

The end result is a CMake error when rerunning the configuration after
non-whitelisted Kconfig warnings.

Fix it by only writing zephyr/.config (and zephyr/include/autoconf.h) in
kconfig.py if there are no warnings-turned-errors.

Also check if the saved checksum file exists in kconfig.cmake before
trying to open it. Normally this shouldn't happen though.

ulfalizer commented:
CC @SebastianBoe

SebastianBoe commented:
Commands should not write to their output file when they have errors.

Could we rework kconfig.py/kconfiglib.py to allow this?

ulfalizer committed:
cmake: kconfig: Fix rerunning cmake after Kconfig warnings

Commit b3d165f ("scripts: kconfig: Handle warnings generated
during evaluation") made it common for kconfig.py to fail after writing
zephyr/.config. This confuses the configuration fragment checksum logic
in cmake/kconfig.cmake, because it expects the saved checksum file to
exist if zephyr/.config exists.

The end result is a CMake error when rerunning the configuration after
non-whitelisted Kconfig warnings.

Fix it by only writing zephyr/.config (and zephyr/include/autoconf.h) in
kconfig.py if there are no warnings-turned-errors.

Also check if the saved checksum file exists in kconfig.cmake before
trying to open it. Normally this shouldn't happen though.

Move the writing of the checksum file to before writing zephyr/.config
as well. That way, zephyr/.config only gets written if the other
operations succeed.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
@SebastianBoe 
Changed

Would be a bit more flexible if some Kconfiglib functions accepted a `file` object...

ulfalizer commented on cmake/kconfig.cmake:
```diff
@@ -103,20 +103,20 @@ endforeach()
-else()
+set(CREATE_NEW_DOTCONFIG 1)
+# Check if the checksum file exists too before trying to open it, though it
+# should under normal circumstances
```
Would happen if something goes wrong between writing `zephyr/.config` and calculating the checksum. Probably best to just retry in that case.

SebastianBoe replied:
Calculating the checksum is done before writing zephyr/.config AFAICT. So this condition would indicate an internal error / unreachable code situation. Internal errors / unreachable code should cause a failure as close to the root as possible to ease debugging (fail-fast).

ulfalizer replied:
It is written after writing `zephyr/.config` though. Another fix would be to write the checksum file before writing `zephyr/.config` I think, so that `zephyr/.config` marks stuff as done.

ulfalizer replied:
I added on that part as well.

SebastianBoe replied:
"It is written after writing zephyr/.config though."

My mistake.
